template.md

```diff
@@ -0,0 +1,139 @@
+# Repository Template
+
+Many of our projects are based on a [repository template](https://github.com/ethereum/ethereum-python-project-template/).
+
+A key benefit of using the template is that we can easily propagate best practices to a bunch of repos.
+
+## Using template to create a repo
+
+Basically, just clone the template repo and start any new repo with that clone.
+
+TODO: step-by-step instructions
+
+## Upgrading a repo to use the latest templated features
+
+### Confirm template is remote
+
+Check that you have the remote:
+
+```sh
+git remote -v
+```
+
+#### What it looks like if you DO have the template remote
+
+```
+$ git remote -v
+origin	git@github.com:carver/eth-typing.git (fetch)
+origin	git@github.com:carver/eth-typing.git (push)
+template-host	git@github.com:ethereum/ethereum-python-project-template.git (fetch)
+template-host	git@github.com:ethereum/ethereum-python-project-template.git (push)
+upstream	git@github.com:ethereum/eth-typing.git (fetch)
+upstream	git@github.com:ethereum/eth-typing.git (push)
+```
+
+#### What it looks like if you do NOT have the template remote
+
+```
+$ git remote -v
+origin	git@github.com:carver/eth-typing.git (fetch)
+origin	git@github.com:carver/eth-typing.git (push)
+upstream	git@github.com:ethereum/eth-typing.git (fetch)
+upstream	git@github.com:ethereum/eth-typing.git (push)
+```
+
+In this case, add the template as a remote this way:
+
+```sh
+git remote add template-host git@github.com:ethereum/ethereum-python-project-template.git
+```
+
+Then check out the master of the template repo as a local branch called `template`:
+
+```sh
+git fetch template-host
+git checkout --track -b template template-host/master
+```
+
+### Pull in the latest template changes
+
+#### Get the latest template changes into your template branch
+
+```sh
+git checkout template
+git pull
+```
+
+#### Create a template upgrade feature branch
+
+```sh
+git checkout master
+git pull
+git checkout -b upgrade-template
+```
+
+#### Merge the latest template changes into your feature branch
+
+```sh
+git merge template
```

Comment:
Thanks for the write up! I tried to follow it closely [in this PR](https://github.com/ethereum/lahja/pull/161) and it *mostly* works but there's a catch.

When I clones this repo as a launchpad for the lahja repo, I squashed all the old history into a single initial commit. So, from a semantic point of view, there's no shared history. The same is true for dopple where I [synced](https://github.com/ethereum/dopple/pull/4) the existing repo with the contents of the template repo.

In those cases this `git merge template` step will fail with:

>`fatal: refusing to merge unrelated histories`

You can still get there though with this slightly different flow from here:

1. First find the first commit from the template repo that can be considered new/needed
2. Then run: `git rebase --onto upgrade-template <first-new-commit-from-template> <latest-commit-from-template>`
3. Work through conflicts interactively just like with any rebase
4. At the end of the process, you are left on a detached head, the changes aren't actually applied to the `upgrade-template` branch which means that as a final step one still needs to create a branch on the current head.

I usually try to stay away from the `--onto` syntax and just use `git cherry-pick <first-commit>..<last-commit>` instead but it turns out that this doesn't work when there are merge commits within that range so the `--onto` based syntax is the only way I've found to make that work in this scenario.

This may actually be a pro argument to just keep the history when forking the template repo but so far I've always considered it cleaner to not carry this *unrelated* (from a projects POV) history with me when I fork the template repo to create a new project.

Otherwise, I think the instructions are fine :+1:

Reply:
> This may actually be a pro argument to just keep the history when forking the template repo but so far I've always considered it cleaner to not carry this _unrelated_ (from a projects POV) history with me when I fork the template repo to create a new project.

Yeah, I think I agree with everything here: all else equal it would be cleaner to squash the template history. But merging is so much cleaner with shared history, and I'd like to have it be as easy as possible to pull in updates from the template to all our repos. So I think that's a strong argument to make new repos by forking the template and keeping its history.

Reply:
>So I think that's a strong argument to make new repos by forking the template and keeping its history.

Yes, I agree. Will do that in the future.